models/user.js

```diff
@@ -8,43 +9,61 @@ const bcrypt = require('bcryptjs');
 const uniqueValidator = require('mongoose-unique-validator');
 
 const UserSchema = mongoose.Schema({
-  createdAt              : { type: Date  },
-  updatedAt              : { type: Date  },
+  createdAt              : { type: Date, default: Date.now },
+  updatedAt              : { type: Date },
   password               : { type: String, select: false },
   username               : { type: String, required: true, unique: true },
   recipesInCart          : { type: Array },
   arrayOfFavoriteRecipes : { type: Array },
 });
 
 // Make sure that only 1 user can exist with the same username
-UserSchema.plugin(uniqueValidator);
+UserSchema.plugin(uniqueValidator, { message: 'Error, expected {PATH} to be unique.' });
 
 // Defines the callback with a regular function to avoid problems with this schema
 UserSchema.pre('save', function (next) {
-  // SET createdAt AND updatedAt
-  const now = new Date();
-  this.updatedAt = now;
-  if (!this.createdAt) {
-    this.createdAt = now;
-  }
-  // ENCRYPT PASSWORD
   const user = this;
-  if (!user.isModified('password')) {
-    return next();
-  }
-  bcrypt.genSalt(10, (err, salt) => {
-    bcrypt.hash(user.password, salt, (errHashing, hash) => {
-      if (errHashing) return next(errHashing);
-      user.password = hash;
-      next();
-    });
+  user.updatedAt = Date.now;
+
+  // if (!user.isModified('password')) {
+  //   return next();
+  // }
+
+  // ENCRYPT PASSWORD
+  bcrypt.hash(user.password, 10, (errHashing, hash) => {
+    if (errHashing) return next(errHashing);
+    user.password = hash;
+    next();
   });
 });
 
-UserSchema.methods.comparePassword = function (password, done) {
-  bcrypt.compare(password, this.password, (err, isMatch) => {
-    done(err, isMatch);
-  });
+UserSchema.statics.authenticate = function (username, password, next) {
+  // eslint-disable-next-line quotes
+  console.log("username:", username, "password:", password);
+  User
+    .findOne({
+      username,
+    }, 'username password') // fetch username and password
+    .then((user) => {
+      if (!user) {
+        const error = new Error('User not found.');
+        error.status = 401;
+        return next(error);
+      }
+      if (user) {
+        // User found, compare password
+        console.log(user);
+        bcrypt.compare(password, user.password, (error, isMatch) => {
```

Comment:
where does isMatch come from?

Reply:
its "taco", can be called anything. It is the result of bcrypt.compare() and is either true or false.

Reply:
👍